opennlp-tools/src/main/java/opennlp/tools/ml/model/OnePassDataIndexer.java

```diff
@@ -39,141 +39,175 @@
-    List<Integer> indexedContext = new ArrayList<>();
-
-    for (int eventIndex = 0; eventIndex < numEvents; eventIndex++) {
-      Event ev = events.removeFirst();
-      String[] econtext = ev.getContext();
-      ComparableEvent ce;
-
-      int ocID;
-      String oc = ev.getOutcome();
-
-      if (omap.containsKey(oc)) {
-        ocID = omap.get(oc);
-      } else {
-        ocID = outcomeCount++;
-        omap.put(oc, ocID);
-      }
-
-      for (String pred : econtext) {
-        if (predicateIndex.containsKey(pred)) {
-          indexedContext.add(predicateIndex.get(pred));
-        }
-      }
-
-      // drop events with no active features
-      if (indexedContext.size() > 0) {
-        int[] cons = new int[indexedContext.size()];
-        for (int ci = 0; ci < cons.length; ci++) {
-          cons[ci] = indexedContext.get(ci);
-        }
-        ce = new ComparableEvent(ocID, cons);
-        eventsToCompare.add(ce);
-      } else {
-        System.err.println("Dropped event " + ev.getOutcome() + ":"
-            + Arrays.asList(ev.getContext()));
-      }
-      // recycle the TIntArrayList
-      indexedContext.clear();
-    }
-    outcomeLabels = toIndexedStringArray(omap);
-    predLabels = toIndexedStringArray(predicateIndex);
-    return eventsToCompare;
-  }
+	public OnePassDataIndexer(){}
+
+	@Override
+	public void index(ObjectStream<Event> eventStream) throws IOException{
+		int cutoff=parameters.getIntParam(CUTOFF_PARAM, CUTOFF_DEFAULT);
+		boolean sort=parameters.getBooleanParam(SORT_PARAM, SORT_DEFAULT);
+		
+		Map<String, Integer> predicateIndex = new HashMap<>();
+		LinkedList<Event> events;
+		List<ComparableEvent> eventsToCompare;
+
+		System.out.println("Indexing events using cutoff of " + cutoff + "\n");
+
+		System.out.print("\tComputing event counts...  ");
+		events = computeEventCounts(eventStream, predicateIndex, cutoff);
+		System.out.println("done. " + events.size() + " events");
+
+		System.out.print("\tIndexing...  ");
+		eventsToCompare = index(events, predicateIndex);
+		// done with event list
+		events = null;
+		// done with predicates
+		predicateIndex = null;
+
+		System.out.println("done.");
+
+		System.out.print("Sorting and merging events... ");
+		sortAndMerge(eventsToCompare, sort);
+		System.out.println("Done indexing.");
+	}
+
+	/**
+	 * Reads events from <tt>eventStream</tt> into a linked list. The predicates
+	 * associated with each event are counted and any which occur at least
+	 * <tt>cutoff</tt> times are added to the <tt>predicatesInOut</tt> map along
+	 * with a unique integer index.
+	 *
+	 * @param eventStream
+	 *          an <code>EventStream</code> value
+	 * @param predicatesInOut
+	 *          a <code>TObjectIntHashMap</code> value
+	 * @param cutoff
+	 *          an <code>int</code> value
+	 * @return a <code>TLinkedList</code> value
+	 */
+	private LinkedList<Event> computeEventCounts(ObjectStream<Event> eventStream,
```

Comment:
Not a problem, I did not change this method, but I will have it return a List<Event>.